ocaml/typing/ctype.ml

```diff
@@ -2158,28 +2195,94 @@ let type_jkind_sub env ty jkind =
      ensure it's a valid argument to [t].  (We believe there are still loops
      like this that can occur, though, and may need a more principled solution
      later).  *)
-  let rec loop ty fuel =
+  let rec expand_and_check_one fuel ty jkind =
     (* This is an optimization to avoid unboxing if we can tell the constraint
        is satisfied from the type_kind *)
     match get_desc ty with
     | Tconstr(p, _args, _abbrev) ->
-        let jkind_bound =
+        let jkind_estimate =
           try (Env.find_type p env).type_jkind
           with Not_found -> Jkind.Primitive.any ~why:(Missing_cmi p)
         in
-        if Jkind.sub jkind_bound jkind
-        then Success
-        else if fuel < 0 then Failure jkind_bound
+        if Jkind.sub jkind_estimate jkind
+        then jkind_estimate, Success
+        else if fuel < 0 then
+          jkind_estimate,
+          Failure { estimate = jkind_estimate; bound = jkind; ty }
         else begin match unbox_once env ty with
-          | Final_result ty -> shallow_check ty
-          | Stepped ty -> loop ty (fuel - 1)
+          | Final_result ty -> check_one fuel ty jkind
+          | Stepped ty -> expand_and_check_one (fuel - 1) ty jkind
           | Missing missing_cmi_for ->
-            Missing_cmi (jkind_bound, missing_cmi_for)
+            jkind_estimate, Missing_cmi missing_cmi_for
         end
-    | Tpoly (ty, _) -> loop ty fuel
-    | _ -> shallow_check ty
+    | Tpoly (ty, _) -> expand_and_check_one fuel ty jkind
+    | _ -> check_one fuel ty jkind
+
+  (* Unlike [expand_and_check_one], this will not expand the type it is passed.
+     But, if that type turns out to be an unboxed product we may end up
+     expanding its components. *)
+  and check_one fuel ty jkind =
+    match estimate_type_jkind_head env ty with
+    | Jkind ty_jkind ->
+      let result =
+        if Jkind.sub ty_jkind jkind then Success else
+          Failure {estimate = ty_jkind; bound = jkind; ty}
+      in
+      ty_jkind, result
+    | TyVar (ty, ty_jkind) ->
+      ty_jkind, Type_vars [{ estimate = ty_jkind; bound = jkind; ty }]
+    | Product component_types as product ->
+      let jkinds =
+        Jkind.is_nary_product (List.length component_types) jkind
+      in
+      match jkinds with
+      | None ->
+        let estimate =
+          jkind_of_result env ~expand_components:(fun x -> x) product
+        in
+        estimate, Failure { estimate; bound = jkind; ty }
+      | Some jkinds ->
+        (* Note: here we "duplicate" the fuel, which may seem like
+           cheating. Fuel counts expansions, and its purpose is to guard against
+           infinitely expanding a recursive type. In a wide product, we many
+           need to expand many types shallowly, and that's fine. *)
+        let results = List.map2 (check_one fuel) component_types jkinds in
+        process_product_results fuel [] [] results
+
+  (* Here, [results] is a list of results from attempting to constrain a product
```

Comment:
I got thrown off by this comment, because `results` is only the unprocessed results. The others are in the accumulators. But as stated, I was confused to see something meaningfully happen when `results` is empty (but of course it's just because that happens at the end of the recursion).